File: java/array/11.ContainerWithMostWater.java
class Solution {
    public int maxArea(int[] height) {
        int left = 0;
        int right = height.length - 1;
        int maxCapacity = 0;

        while(left < right) {
            int width = right - left;
            int minHeight = Math.min(height[left], height[right]);
            int currentCapacity = width * minHeight;
            
            maxCapacity = Math.max(currentCapacity, maxCapacity);

            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }
        return maxCapacity;
    }
}

File: python/array/11.container_with_most_water.py

# Link --> https://leetcode.com/problems/container-with-most-water/

def max_area(height):
    left = 0
    right = len(height) - 1
    max_capacity = 0

    while left < right:
        width = right - left
        min_height = min(height[left], height[right])
        current_capacity = min_height * width
        max_capacity = max(current_capacity, max_capacity)

        if (height[left] < height[right]):
            left += 1
        else:
            right -= 1
    
    return max_capacity
    

height = [1, 8, 6, 2, 5, 4, 8, 3, 7]
print(max_area(height))

